internal/xdscache/v3/runtime.go

```diff
@@ -20,30 +20,46 @@ import (
 	envoy_v3 "github.com/projectcontour/contour/internal/envoy/v3"
 	"github.com/projectcontour/contour/internal/protobuf"
 	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/structpb"
 )
 
+type ConfigurableRuntimeSettings struct {
+	MaxRequestsPerIOCycle *uint32
+}
+
 // RuntimeCache manages the contents of the gRPC RTDS cache.
-type RuntimeCache struct {
+type runtimeCache struct {
 	contour.Cond
+	runtimeKV map[string]*structpb.Value
+}
+
+// NewRuntimeCache builds a RuntimeCache with the provided runtime
+// settings that will be set in the runtime layer configured by Contour.
+func NewRuntimeCache(runtimeSettings ConfigurableRuntimeSettings) *runtimeCache {
+	runtimeKV := make(map[string]*structpb.Value)
+	if runtimeSettings.MaxRequestsPerIOCycle != nil && *runtimeSettings.MaxRequestsPerIOCycle > 0 {
+		runtimeKV["http.max_requests_per_io_cycle"] = structpb.NewNumberValue(float64(*runtimeSettings.MaxRequestsPerIOCycle))
+	}
+	return &runtimeCache{runtimeKV: runtimeKV}
 }
```

Comment:
Question: would it make sense to move the existing (fixed) `re2.max_program_size.error_level` and `re2.max_program_size.warn_level` runtime parameters from `internal/envoy/v3/runtime.go` and set them to `runtimeKV` here as well? In that case all the runtime keys could be found in single file.

Reply:
At the moment we also use this helper from `internal/envoy/v3` to build up the bootstrap config base RTDS layer: https://github.com/projectcontour/contour/blob/f52ad069b1090aaabf8e05866af4ef4e1e4bdb31/internal/envoy/v3/bootstrap.go#L169

We may be able to get rid of that static bootstrap runtime layer since it was introduced in https://github.com/projectcontour/contour/pull/4379 and just serve dynamic layers instead (added in https://github.com/projectcontour/contour/pull/4387) which would mean we can refactor and consolidate a bit.

I think I would prefer to do that refactor in a separate commit/PR though

Reply:
> At the moment we also use this helper from internal/envoy/v3 to build up the bootstrap config

Ah right, that makes sense. It seemed unnecessarily complicated, when not realizing that :smile:

Reply:
added https://github.com/projectcontour/contour/issues/5845 as a cleanup issue